### ISSUE TITLE
FR - Spanner - ID Generation for UUIDs

### ISSUE THREAD

**Issue by saturnism** [spring-cloud/spring-cloud-gcp#857]
In Spanner, support ID Generation, based on annotation.

I don't feel we can get into full JPA-type ID generation. However, if the type is a String or UUID, we can consider automatically generate a new UUID if annotated w/ a UUID strategy.

**Comment by meltsufin** [comment]
Would it make sense to just automatically generate the ID when one is not specified, without requiring an annotation requesting it?

**Comment by ChengyuanZhao** [comment]
@meltsufin what do you mean specifically? 

When there are not any  properties labeled as ID? 
When there is a prop labeled as ID but is null?
When there is a prob labeled as ID and is null but not labeled as auto-generate?
What about if there are multiple properties that comprise the PK? What if some of them are null but others aren't?

**Comment by meltsufin** [comment]
Since this is about UUIDs, I assume it implies non-composite IDs.
So from your list, if there is a (String?) property annotated with `@ID` but null.

**Comment by meltsufin** [comment]
In that case, I would assume that the parent foreign key is not something we can generate, but has to be specified. So we would only auto-generate the components that are `null`. Would that make sense?

**Comment by ChengyuanZhao** [comment]
I agree we definitely can't auto-generate the key components that are from the parent, but I have a different idea of the policy we should have. 

My proposal is to have a boolean attribute in our key component annotation called AutoGenerateOnWrite ,so requiring the user to explicitly mark the key components for auto-generated UUID. 

The pros are:
- if the user marks a parent column for auto-gen, then he will get back an error from the client lib during a write attempt saying that parent ID doesn't exist.
- also, it means we don't need to do schema/metadata checking for the user to tell him if a column is a parent key column. This also means a given child class doesn't need to be linked in Java with a parent class, because such relationships don't need to exist from one spanner instance to another. 
- the specific name "AutoGenerateOnWrite" makes it clear to the user exactly when the auto-gen happens. While we will certainly also explain this in the docs, it's a good reminder and reduces confusion.
- the user also doesn't need to make assumptions or guesses about what types of checking we might be doing on his behalf. the behavior becomes more clear
- we give him the option of crashing on null key values. he may not want us to auto-gen in every situation.

**Comment by saturnism** [comment]
In JPA, there is "Generated" annotation that specifies the different strategy.  Would probably need a more detailed design doc and chat w/ Oliver.
